app/services/wallets/balance/decrease_service.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module Wallets
+  module Balance
+    class DecreaseService < BaseService
+      def initialize(wallet:, credits_amount:)
+        super(nil)
+
+        @wallet = wallet
+        @credits_amount = credits_amount
+      end
+
+      def call
+        amount = wallet.rate_amount * credits_amount
+
+        wallet.update!(
+          balance: wallet.balance - amount,
+          credits_balance: wallet.credits_balance - credits_amount,
+          last_balance_sync_at: Time.zone.now,
+          consumed_credits: wallet.consumed_credits + credits_amount,
+          consumed_amount: wallet.consumed_amount + amount,
+          last_consumed_credit_at: Time.current
+        )
+      end
```

Comment:
Maybe we should return a result here. What do you think?

```ruby
  result.wallet = wallet
  result
rescue ActiveRecord::RecordInvalid => e
  result.fail_with_validations!(e.record)
```

Reply:
wallet is actually not used in a caller for now, but added it if that's the standard :)